File: datamart/src/main/scala/DataMart.scala
import clickhouse.{Reader, Writer}
import feature.Preprocessor
import org.apache.spark.sql.{DataFrame, SparkSession}

object DataMart {
  private val CLICKHOUSE_URL = "jdbc:clickhouse://clickhouse"
  private val USER = "default"
  private val PASSWORD = ""
  private val DRIVER = "com.clickhouse.jdbc.ClickHouseDriver"
  private val OPEN_FOOD_TABLE = "datasets.openfood"
  private val OPEN_FOOD_PREDICTIONS_TABLE = "datasets.openfood_predictions"

  private def readOpenFoodDataset(): DataFrame = {
    implicit val spark: SparkSession = SparkSession.builder.getOrCreate()
    val reader = Reader(CLICKHOUSE_URL, USER, PASSWORD, DRIVER, OPEN_FOOD_TABLE)
    reader.read(spark)
  }

  def readTransformOpenFoodDataset(): DataFrame = {
    val rawDataset = readOpenFoodDataset()

    val transforms: Seq[DataFrame => DataFrame] = Seq(
      Preprocessor.toFloat,
      Preprocessor.fillNull,
      Preprocessor.toVector,
      Preprocessor.scale
    )

    val transformed = transforms.foldLeft(rawDataset) { (df, f) => f(df) }

    transformed
  }

  def writeClickhouse(df: DataFrame): Unit = {
    val writer = Writer(CLICKHOUSE_URL, USER, PASSWORD, DRIVER, OPEN_FOOD_PREDICTIONS_TABLE)
    writer.write(df)
  }
}

File: src/datamart.py
from pyspark.sql import DataFrame, SparkSession, SQLContext


class DataMart:
    def __init__(self, spark: SparkSession):
        self.spark_context = spark.sparkContext
        self.sql_context = SQLContext(self.spark_context, spark)
        self.jwm_datamart = self.spark_context._jvm.DataMart

    def load_dataset(self) -> DataFrame:
        jvm_data = self.jwm_datamart.readTransformOpenFoodDataset()
        return DataFrame(jvm_data, self.sql_context)

    def write_predictions(self, df: DataFrame):
        self.jwm_datamart.writeClickhouse(df._jdf)
